domain/user/service/service.go

```diff
@@ -170,6 +175,35 @@ func (s *Service) GetUserByName(
 	return usr, nil
 }
 
+// GetUserWithAuth will find and return the user with UUID. If there is no
+// user for the UUID then an error that satisfies usererrors.NotFound will
+// be returned.
+//
+// GetUserWithAuth will not return users that have been previously removed.
+func (s *Service) GetUserWithAuth(
+	ctx context.Context,
+	uuid user.UUID,
+	password string,
+) (user.User, error) {
+	if err := uuid.Validate(); err != nil {
+		return user.User{}, errors.Annotatef(usererrors.UUIDNotValid, "validating uuid %q", uuid)
+	}
+
+	usr, err := s.st.GetUserWithAuth(ctx, uuid)
+	if err != nil {
+		return user.User{}, errors.Annotatef(err, "getting user for uuid %q", uuid)
+	}
+
+	passwordHash, err := auth.HashPassword(auth.NewPassword(password), usr.PasswordSalt)
+	if err != nil {
+		return user.User{}, errors.Annotatef(err, "hashing password for user with uuid %q", uuid)
+	} else if passwordHash != usr.PasswordHash {
+		return user.User{}, errors.Annotatef(usererrors.Unauthorized, "invalid password for user with uuid %q", uuid)
+	}
```

Comment:
The hash of the password should be passed into `GetUserWithAuth`. We do it this way, to ensure that the users password can't change between the `GetUserWithAuth` and the checking of the password.

It's run inside of the transaction.

Reply:
done